### PR TITLE
Connect the individual prediction gradients into the Combination workflow

### PR DESCRIPTION
**Pull request by kaminow**
Currently the `Combination` implementations only return the final prediction, so if we want to calculate a loss against the individual pose predictions, we get an error since the computation graph for the individual predictions are empty. Want to connect in the individual pose prediction gradients to allow using a loss against these pose predictions (eg cross entropy loss for selecting the best pose).

### CODE REVIEW

**Comment by kaminow** [comment]
this might just be as simple as returning the individual pose predictions in the `Combination.forward` methods and then returning their gradients in the `Combination.backward` methods

File: mtenn/combination.py [hunk 1]
```diff
@@ -226,11 +226,15 @@ def backward(ctx, grad_output):
...
         for n, grad_list in grad_dict.items():
-            final_grads[n] = torch.stack(grad_list, axis=-1).mean(axis=-1)
+            # Gradient from final prediction loss
+            cur_final_grad = comb_grad * torch.stack(grad_list, axis=-1).mean(axis=-1)
 
-        # Adjust gradients by grad_output
-        for grad in final_grads.values():
-            grad *= grad_output
+            # Add in gradients for each pose from per-pose loss
+            for pose_grad, param_grad in zip(pose_grads, grad_list):
```
Comment (hmacdope) [inline mtenn/combination.py:233]: Add a check for lengths.

File: mtenn/combination.py [hunk 2]
```diff
@@ -409,9 +428,12 @@ def backward(ctx, grad_output):
                 .sum(axis=-1)
             )
 
-        # Adjust gradients by grad_output
-        for grad in final_grads.values():
-            grad *= grad_output
+            # Add in gradients for each pose from per-pose loss
+            for pose_grad, param_grad in zip(pose_grads, grad_list):
```
Comment (hmacdope) [inline mtenn/combination.py:432]: Check for lengths.

**Review by hmacdope — APPROVED** [review]
Minor nitpick, looks good otherwise. Can't vouch for math though.

**Comment by kaminow** [comment]
@jchodera @hmacdope I think I'm pretty happy with the level of commenting/documentation now, but please take a look when you get a chance and let me know what you think!

**Review by hmacdope — APPROVED** [review]
Looks good to me, nicely detailed and much easier to follow.
